rclpy/src/rclpy/_rclpy_qos_event.c

```diff
@@ -0,0 +1,357 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <rcl/event.h>
+
+typedef union _qos_event_callback_data {
+  // Subscription events
+  rmw_requested_deadline_missed_status_t requested_deadline_missed;
+  rmw_liveliness_changed_status_t liveliness_changed;
+  // Publisher events
+  rmw_offered_deadline_missed_status_t offered_deadline_missed;
+  rmw_liveliness_lost_status_t liveliness_lost;
+} _qos_event_callback_data_t;
+
+typedef PyObject * (* _qos_event_data_filler_function)(_qos_event_callback_data_t *);
+
+static
+bool
+_check_rcl_return(rcl_ret_t ret, const char * error_msg)
+{
+  if (RCL_RET_OK != ret) {
+    PyErr_Format(PyExc_RuntimeError,
+      "%s: %s", error_msg, rcl_get_error_string().str);
+    rcl_reset_error();
+    return false;
+  }
+  return true;
+}
+
+static
+void
+_destroy_event_capsule(PyObject * pycapsule)
+{
+  rcl_event_t * event = (rcl_event_t *)PyCapsule_GetPointer(pycapsule, "rcl_event_t");
+  if (!event) {
+    PyErr_Clear();
+    int stack_level = 1;
+    PyErr_WarnFormat(
+      PyExc_RuntimeWarning, stack_level, "_destroy_event_capsule failed to get pointer");
+  }
+  rcl_ret_t ret = rcl_event_fini(event);
+  if (RCL_RET_OK != ret) {
+    int stack_level = 1;
+    PyErr_WarnFormat(
+      PyExc_RuntimeWarning, stack_level, "Failed to fini event: %s", rcl_get_error_string().str);
+  }
+  PyMem_Free(event);
+}
+
+static
+bool
+_is_pycapsule_rcl_subscription(PyObject * pycapsule)
+{
+  return PyCapsule_IsValid(pycapsule, "rclpy_subscription_t");
+}
+
+static
+bool
+_is_pycapsule_rcl_publisher(PyObject * pycapsule)
+{
+  return PyCapsule_IsValid(pycapsule, "rclpy_publisher_t");
+}
+
+static
+rcl_subscription_t *
+_pycapsule_to_rcl_subscription(PyObject * pycapsule)
+{
+  return (rcl_subscription_t *)PyCapsule_GetPointer(pycapsule, "rclpy_subscription_t");
+}
+
+static
+rcl_publisher_t *
+_pycapsule_to_rcl_publisher(PyObject * pycapsule)
+{
+  return (rcl_publisher_t *)PyCapsule_GetPointer(pycapsule, "rclpy_publisher_t");
```

Comment:
Just noting that this works because the first member of the `rclpy_publisher_t` struct is an `rcl_publisher_t`. It would be more explicit to use a temporary and return the address of the publisher in that struct instead.

Reply:
Noted - I was aware of that - you're right that it would be more explicit to do it the other way. I'd like to do a bit of refactoring after this, but for this change let's leave it the way it is.

Reply:
Since this needs another round of CI to check the windows warning, mind updating this one before that?

Reply:
Ok - I am just running the tests locally first - and I have updated the usage from the other PR as well - should be just a minute

Reply:
@sloretz OK - i've updated the code and it's looking good locally - the same repos list gist is accurate https://gist.githubusercontent.com/emersonknapp/67d93f9e90183987832715b5979640e8/raw/7eb11515e048841d603d35ce0a5e6fecfdf9296f/qos-4-callbacks.repos